dali/pipeline/operator/sequence_operator.h

```diff
@@ -0,0 +1,627 @@
+    const auto &input = ws.template Input<InputBackend>(input_idx);
+    auto sample_dim = input.shape().sample_dim();
+    DALI_ENFORCE(
+        sample_dim > num_expand_dims,
+        make_string("Cannot flatten the sequence-like input ", input_idx,
+                    ". Samples must have more dimensions (got ", sample_dim,
+                    ") than the requested number of dimensions to unfold: ", num_expand_dims, "."));
+    // TODO(ktokarski) TODO(klecki)
+    // Rework it when TensorList stops being contigious and supports "true sample" mode
+    auto expanded_input = unfold_outer_dims(input, num_expand_dims);
+    const auto &input_layout = input.GetLayout();
+    expanded_input.SetLayout(input_layout.sub(num_expand_dims));
+    return std::make_shared<typename ws_input_t<InputBackend>::element_type>(
+        std::move(expanded_input));
+  }
+
+  template <typename OutputBackend>
+  ws_input_t<OutputBackend> ExpandOutput(const workspace_t<Backend> &ws, int output_idx,
+                                         int num_expand_dims) {
+    const auto &output = ws.template Output<OutputBackend>(output_idx);
+    auto sample_dim = output.shape().sample_dim();
+    DALI_ENFORCE(
+        sample_dim > num_expand_dims,
+        make_string("Cannot flatten the sequence-like output ", output_idx,
+                    ". Samples must have more dimensions (got ", sample_dim,
+                    ") than the requested number of dimensions to unfold: ", num_expand_dims, "."));
+    // TODO(ktokarski) TODO(klecki)
+    // Rework it when TensorList stops being contigious and supports "true sample" mode
+    auto expanded_output = unfold_outer_dims(output, num_expand_dims);
+    return std::make_shared<typename ws_input_t<OutputBackend>::element_type>(
+        std::move(expanded_output));
+  }
+
+  template <typename OutputBackend>
+  void SetOutputLayout(const workspace_t<Backend> &ws, int output_idx, TensorLayout layout_prefix) {
+    DALI_ENFORCE(ExpandedOutputIsType<OutputBackend>(output_idx));
+    auto &expanded_output = ExpandedOutput<OutputBackend>(output_idx);
+    auto &output = ws.template Output<OutputBackend>(output_idx);
+    const auto &layout = expanded_output.GetLayout();
+    if (layout.size() == 0) {
+      output.SetLayout(layout);
+    } else {
+      auto expanded_layout = layout_prefix + layout;
+      output.SetLayout(expanded_layout);
+    }
+  }
+
+  void SetupExpandedWorkspace(workspace_t<CPUBackend> &expanded,
+                              const workspace_t<CPUBackend> &ws) {
+    expanded.SetThreadPool(&ws.GetThreadPool());
+  }
+
+  void SetupExpandedWorkspace(workspace_t<GPUBackend> &expanded,
+                              const workspace_t<GPUBackend> &ws) {
+    if (ws.has_stream()) {
+      expanded.set_stream(ws.stream());
+    }
+  }
+
+  template <typename InputBackend>
+  void AddInputHelper(ws_input_t<InputBackend> input) {
+    expanded_.AddInput(input);
+  }
```

Comment:
Actually, this "helper" hurts, not helps - AddInput handles different backends gracefully (or so it seems) whereas AddInputHelper requires explicit template argument.